File: code/example_1.cpp
#include <kqp/feature_matrix/dense.hpp>
#include <kqp/kernel_evd/incremental.hpp>
#include <kqp/probabilities.hpp>

int main(int, const char**) {
    
    // --- Compute a density at random
    
    // Definitions
    using namespace kqp;
    typedef Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic> Matrix;
    int dim = 10;
    
    // Creating an incremental builder
    IncrementalKernelEVD<DenseMatrix<double>> kevd;
    
    // Add 10 vectors with $\alpha_i=1$
    for(int i = 0; i < 10; i++) {
        // Adds a random $\varphi_i$
        Matrix m = Matrix::Random(dim, 1);
        kevd.add(DenseMatrix<double>(m));
    }
    
    // Get the result $\rho \approx X Y D Y^\dagger X^\dagger$
    DenseMatrix<double> mX;
    typename AltDense<double>::type mY;
    Eigen::Matrix<double, Eigen::Dynamic,1>  mD;
    
    kevd.get_decomposition(mX, mY, mD);

    // --- Compute a kEVD for a subspace
    
    IncrementalKernelEVD<DenseMatrix<double>> kevd_event;
    for(int i = 0; i < 3; i++) {
        // Adds a random $\varphi_i$
        Matrix m = Matrix::Random(dim, 1);
        kevd_event.add(DenseMatrix<double>(m));
    }

    
    // --- Compute some probabilities
    
    // Setup densities and events
    Density<DenseMatrix<double>> rho(kevd);
    Event<DenseMatrix<double>> event(kevd_event);
    
    // Compute the probability
    std::cout << "Probability = " << rho.probability(event) << std::endl;

    // Conditional probability
    Density<DenseMatrix<double>> rho_cond = event.project(kevd).normalize(); 

    // Conditional probability (orthogonal event)
    Density<DenseMatrix<double>> rho_cond_orth = event.project(kevd, true).normalize();
    
    return 0;
}

File: examples/example_1.java
package net.bpiwowar.qia.tasks; 
/**
 * @author B. Piwowarski <benjamin@bpiwowar.net>
 * @date 19/4/12
 */
import net.bpiwowar.kqp.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URL;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.Map;
import java.util.Properties;

public class TestKQP {
  
   static public void main(String [] args) {
        int dim = 10;

        // Creating an incremental builder
        System.err.format("Creating a KEVD builder%n");
        KEVDAccumulatorDenseDouble kevd = new KEVDAccumulatorDenseDouble();

        // Add 10 vectors with $\alpha_i=1$
        System.err.format("Adding 10 vectors%n");
        for(int i = 0; i < 10; i++) {
            // Adds a random $\varphi_i$
            EigenMatrixDouble m = new EigenMatrixDouble(dim,1);
            m.randomize();
            kevd.add(new DenseDouble(m));
        }

        // Get the result $\rho \approx X Y D Y^\dagger X^\dagger$
        System.err.println("Getting the result");
        DecompositionDenseDouble d = kevd.getDecomposition();

        // --- Compute a kEVD for a subspace

        System.err.format("Creating a KEVD builder (event)%n");

        KEVDAccumulatorDenseDouble kevd_event = new KEVDAccumulatorDenseDouble();

        for(int i = 0; i < 3; i++) {
            // Adds a random $\varphi_i$
            EigenMatrixDouble m = new EigenMatrixDouble(dim,1);
            m.randomize();
            kevd_event.add(new DenseDouble(m));
        }


        // --- Compute some probabilities


        // Setup densities and events
        d = kevd.getDecomposition();
        System.err.println("Creating the density rho and event E");
        DensityDenseDouble rho = new DensityDenseDouble(kevd);
        rho.normalize();
        EventDenseDouble event = new EventDenseDouble(kevd_event);
        System.err.println("Computing some probabilities");

        // Compute the probability
        System.out.format("Probability = %g%n", rho.probability(event));

        // Conditional probability
        DensityDenseDouble rho_cond = event.project(rho).normalize();
        System.out.format("Entropy of rho/E = %g%n", rho_cond.entropy());

        // Conditional probability (orthogonal event)
        DensityDenseDouble rho_cond_orth = event.project(rho, true).normalize();
        System.out.format("Entropy of rho/not E = %g%n", rho_cond.entropy());

    }

}
